File: 282. Expression Add Operators/C++/Solution.cpp
class Solution {
public:
    vector<string> addOperators(string num, int target) {
        class Utils {
        public:
            void
            backTrace(string &num, int target, int i, string current, vector<string> &result, long value, long prev) {
                if (i == num.length()) {
                    if (value == target) {
                        result.push_back(current);
                    }
                    return;
                }
                for (int j = i; j < num.length(); j++) {
                    if (j != i && num[i] == '0') {
                        break;
                    }
                    string nowInStr = num.substr(i, j + 1 - i);
                    long nowInInt = stol(nowInStr);
                    if (i == 0) {
                        backTrace(num, target, j + 1, nowInStr, result, nowInInt, nowInInt);
                    } else {
                        backTrace(num, target, j + 1, current + "+" + nowInStr, result, value + nowInInt, nowInInt);
                        backTrace(num, target, j + 1, current + "-" + nowInStr, result, value - nowInInt, -nowInInt);
                        backTrace(num, target, j + 1, current + "*" + nowInStr, result, value - prev + prev * nowInInt,
                                  prev * nowInInt);
                    }
                }
            }
        };
        Utils utils;
        vector<string> result;
        string temp = "";
        utils.backTrace(num, target, 0, temp, result, 0, 0);
        return result;
    }
};

File: 282. Expression Add Operators/Python/Solution.py
class Solution(object):
    def addOperators(self, num, target):
        """
        :type num: str
        :type target: int
        :rtype: List[str]
        """
        def back_trace(num, target, i, current, result, value, prev):
            if i == len(num):
                if value == target:
                    result.append(current)
                return
            for j in range(i, len(num)):
                if j != i and num[i] == '0':
                    break
                now_in_str = num[i:j + 1]
                now_in_int = int(now_in_str)
                if i == 0:
                    back_trace(num, target, j + 1, now_in_str, result, now_in_int, now_in_int)
                else:
                    back_trace(num, target, j + 1, current + '+' + now_in_str,
                               result, value + now_in_int, now_in_int)
                    back_trace(num, target, j + 1, current + '-' + now_in_str,
                               result, value - now_in_int, -now_in_int)
                    back_trace(num, target, j + 1, current + '*' + now_in_str, result, value - prev + prev * now_in_int,
                               prev * now_in_int)

        result = []
        back_trace(num, target, 0, '', result, 0, 0)
        return result
